Authentication doc (#115)
cozy/cozy.github.io: src/references/auth.md

---
title: Authentication
---

## Authentication Sequence

1. The User opens up a browser to check his cozy, using his *cozy app domain* (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...)

> 2. he's redirected to his *cozy user domain* (`user.mycozy.cloud`)

Reviewer: ```suggestion
2. (s)he's redirected to his *cozy user domain* (`user.mycozy.cloud`)
```

> 3. The browser request the *cozy user domain*, which send back the authentication form. The user fills his password, sending it back to the  *cozy user domain*

Reviewer: ```suggestion
3. The browser requests the stack through *cozy user domain*, which sends back the authentication form. The user fills his password, sending it back to the  *cozy user domain*
```

> 4. The server *cozy user domain* set an auth cookie for *cozy user domain* ONLY

Reviewer: ```suggestion
4. The server *cozy user domain* sets an auth cookie for *cozy user domain* ONLY
```

Why ONLY ?

Reviewer 2: This cookie is restricted to this sub-domain (and httpOnly). It will *not* be seen by the app domain.

This is by design, so that an app could not use the cookie and authenticate with another app permission.

Reviewer 3: It's a bit confused. Let's rephrase it:

- we are using `httpOnly` to ensure that the JS can't play with the cookie
- the fact that we have distinct cookies for the user domain and the app domain is not by design (for self-hosted, it is the same cookie that is used)
- a cookie on `alice.mycozy.cloud` is not used by `alice-drive.mycozy.cloud` (it's how the cookies work).

We can put the cookie on level upper, on `.mycozy.cloud`, but it has some issues (a user can't be logged on two cozy instances at the same time, the cookies are sent when a user visit a page shared by link on another instance, etc.). So, it was simpler to use several cookies with a copy mechanism than to deal with those issues.

5. A redirect status response code (302) for the requested app domain (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...) is sent to the browser

> 5. The browser request the *cozy app domain* with a [jwt code](https://jwt.io/)

Reviewer: ```suggestion
5. The browser requests the *cozy app domain* with a [jwt code](https://jwt.io/)
```

Reviewer 4: Is this using the `Authorization` header here?

Reviewer 2: Nope. The 302 redirection from user.mycozy.cloud to user-home.mycozy.cloud adds a `?code=xxxxx` to the URL where xxxx is the JWT.

Ok, it's maybe usefull to write it explicitely

Reviewer 4: I think so, it's more clear for me now, thanks 👍

> 5. The browser request the *cozy app domain* with a [jwt code](https://jwt.io/)

Reviewer 3: Well, it's a detail but the code used to transfer the session from the cozy user domain to the cozy app domain is not a JWT (just a random string)

Reviewer 3: The JWT token is not used for transfering the cookie from one domain to another: it's the random code that is used for that. And when the cookie is set on the app domain, the app can fetch its index.html page with the JWT token inside it, and extracts it in JS.

> 6. The jwt code  is checked and decoded through the stack, which send back a bearer token in the request header to *cozy app domain*

Reviewer: ```suggestion
6. The jwt code  is checked and decoded through the stack, which sends back a bearer token in the request header to *cozy app domain*
```

> 6. The jwt code  is checked and decoded through the stack, which send back a bearer token in the request header to *cozy app domain*

Reviewer 3: The stack doesn't send a bearer token in the request header. Well, the stack doesn't sent the request here. And the JWT bearer token is sent in the body of the HTML response. It's then the client that extracts this token and send it to the stack on the *cozy user domain* as a bearer token in the Authorization header of HTTP requests.

Reviewer 2: That was my mistake. Sorry Julie. I thought the code given to from the user domain to the app domain was a JWT and the bearer token was an simple unique string. It appears it's the opposite.

For the client, both string should be used as opaque strings anyway

7. *cozy app domain* uses the bearer token to communicate with the stack through XHR request
8. *cozy app domain* sends back the authenticated UI to the browser, and a new cookie for the *cozy app domain* (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...)

For each new app requested by the user, repeat steps

> (Steps 3 & 4 only happens when the user is not authenticated at all : meaning there is no auth cookie for *cozy user domain* )

Reviewer 3: [nitpick] no space before `:` in english

![Architecture](../img/dev/auth_sequence.png)

## Authentication types

### Password

- Used by the user on the authentication form on the *cozy user domain* (`user.mycozy.cloud`)
- It can be changed through the setting apps

> - It can be reset on the authentication form or in the setting app, by sending an email to the user

Reviewer 3: A fact that may be relevant here: when the password is changed, all the old cookies are invalidated, and we recreate a new one for the current session. In practice, it means the user will need to login again in their other browsers.

### Cookies

- For each app, there is a specific cookie, meaning there is a new cookie for each new domain/sub-domain
- The "Remember me" check box only affect the expiration date of the *cozy user domain* cookie

 !!! warning ""
     - local dev servers (`cozy.tools`), selfHost server and Gandi server use sudomains : `home.user.cozy`
     - dev servers (`cozy.wtf`), integration servers (`cozy.works`), staging servers (`cozy.rocks`) and production server (`mycozy.cloud`) use different domains `user-home.cozy...`

### 2FA

- Can be activated by the user

> - Only works through emails right window

Reviewer 5: I think you wanted to write "right now", no?

> - Once 2FA activated, and the user authenticated, a new chekbox "Trust this machine" create a new cookie (if checked)

Reviewer: ```suggestion
- Once 2FA activated, and the user authenticated, a new checkbox "Trust this machine" creates a new cookie (if checked)
```

 !!! warning ""
     There are a few timing issues with 2FA, as the code validation time is quite short, and some emails may be delayed.

> ### Authentication uses

Reviewer: ```suggestion
### Usecases and their authentication methods
```

7. The jwt token is sent back to the stack, on the *cozy user domain* as a bearer token in the Authorization header of HTTP requests.
8. *cozy app domain* uses the bearer token to communicate with the stack through XHR request
9. *cozy app domain* sends back the authenticated UI to the browser, and a new cookie for the *cozy app domain* (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...)
=======
4. The server *cozy user domain* sets an auth cookie for *cozy user domain* ONLY
5. A redirect status response code (302) for the requested app domain (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...) is sent to the browser
5. The browser requests the *cozy app domain* with a [jwt code](https://jwt.io/)
6. The jwt code  is checked and decoded through the stack, which sends back a bearer token in the request header to *cozy app domain*
7. *cozy app domain* uses the bearer token to communicate with the stack through XHR request
8. *cozy app domain* sends back the authenticated UI to the browser, and a new cookie for the *cozy app domain* (`user-home.mycozy.cloud/`, or `user-drive.mycozy.cloud/`, or ...)

> >>>>>>> d8565b7a7fef9eb08212f0c6877a71f5ab209e1b

Reviewer 6: conflicts to fix
